lib/oauth2-client.js

```diff
@@ -0,0 +1,127 @@
+/*!
+ * Copyright (c) 2021 Digital Bazaar, Inc. All rights reserved.
+ */
+import * as storage from './storage.js';
+import * as bedrock from 'bedrock';
+import {httpClient} from '@digitalbazaar/http-client';
+import {LruCache} from '@digitalbazaar/lru-memoize';
+import logger from './logger.js';
+const {util: {BedrockError}} = bedrock;
+
+let cache;
+
+bedrock.events.on('bedrock.init', async () => {
+  const cfg = bedrock.config['oauth2-client'];
+  cache = new LruCache({
+    max: cfg.cache.maxSize,
+    maxAge: cfg.cache.maxAge
+  });
+});
+
+export class OAuth2Client {
+  /**
+   * Constructs an OAuth2Client instance to return an access token. It first
+   * checks the database to see if an access token already exists for the client
+   * id; if not, it creates an access token then stores it in the database.
+   *
+   * @param {object} options - Options to use.
+   * @param {string} options.config - The "oauth2-client-grant" properties.
+   * @param {object} options.httpsAgent - A NodeJS HTTPS Agent (`https.Agent`).
+   *
+   * @returns {OAuth2Client} The `OAuth2Client` instance.
+   */
+  constructor({config, httpsAgent} = {}) {
+    this.client_id = config.client_id;
+    this.client_secret = config.client_secret;
+    this.token_endpoint = config.token_endpoint;
+    this.grant_type = config.grant_type;
+    this.scope = config.scope;
+    this.maxRetries = config.maxRetries;
+    this.httpsAgent = httpsAgent;
+  }
+
+  /**
+   * Gets a new access token by first checking the database to see if an
+   * access token already exists for the client id; if not it creates an
+   * access token then stores it in the database.
+   *
+   * @returns {string} The access token.
```

Comment:
Should this include `Promise` (here and in any other async functions)?

```suggestion
   * @returns {Promise<string>} The access token.
```